### PR TITLE
Invalidate identity cookie on logout.

### PR DESCRIPTION
**Pull request by mohavee**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ✔️
| Tests pass?   | ✔️

Cycle user auth key on user logout.

### CODE REVIEW

File: framework/web/IdentityInterface.php [hunk 1]
```diff
@@ -102,4 +113,20 @@ public function getAuthKey();
...
     public function validateAuthKey($authKey);
+
+    /**
+     * Update the "auth key" token for the given user identity.
+     *
+     * @param IdentityInterface $identity
+     * @param string $token
+     * @return void
+     */
+    public function updateAuthKey(IdentityInterface $identity, $token);
```
Comment (bizley) [inline framework/web/IdentityInterface.php:124]: We cannot introduce new methods in this interface, it will be major BC break.

**Review by bizley — COMMENTED** [review]
Overall I think this is a good idea but unfortunately I don't see a good way to introduce it in Yii 2 because of all the BC breaking changes. I would recommend to have something like this in Yii 3 and here I would only add an adnotation in [the docs](https://github.com/yiisoft/yii2/blob/master/docs/guide/security-best-practices.md) explaining how to invalidate authKey on logout (maybe using event based implementation).

**Comment by rob006** [comment]
Wouldn't that logout all sessions? So if I'm logged in on 5 devices, clicking "logout" on one of them will invalidate this cookie on all devices?

**Comment by mohavee** [comment]
> Wouldn't that logout all sessions? So if I'm logged in on 5 devices, clicking "logout" on one of them will invalidate this cookie on all devices?

yep, we would need multiple auth keys used per device to support that, at this moment i think is better to have a way to minimize damage after stealing identity cookie rather than support "remember me" feature on all logged in devices.

File: framework/web/IdentityInterface.php (hunk 1 above)
Comment (mohavee) [reply framework/web/IdentityInterface.php:124]: I know in current design i didnt find a way to not break BC

**Comment by rob006** [comment]
> at this moment i think is better to have a way to minimize damage after stealing identity cookie rather than support "remember me" feature on all logged in devices.

Why do you think that logout is equivalent of "stealing identity cookie"? I'm using logout action all the time and never did it because of my identity cookie might be stolen. This level of security/annoyance definitely should not enforced by framework - if you need it you can easily do that on `afterLogout()` (but from UX perspective you probably should implement it as explicit "logout all sessions" action instead triggering it implicitly on regular logout). Especially that changing auth key does not really logout existing sessions (https://github.com/yiisoft/yii2/issues/9718) and it probably can't be fixed at framework level.

**Comment by mohavee** [comment]
> > at this moment i think is better to have a way to minimize damage after stealing identity cookie rather than support "remember me" feature on all logged in devices.
> 
> Why do you think that logout is equivalent of "stealing identity cookie"? I'm using logout action all the time and never did it because of my identity cookie might be stolen. This level of security/annoyance definitely should not enforced by framework - if you need it you can easily do that on `afterLogout()` (but from UX perspective you probably should implement it as explicit "logout all sessions" action instead triggering it implicitly on regular logout). Especially that changing auth key does not really logout existing sessions (#9718) and it probably can't be fixed at framework level.

i dont think its equivalent, just saying that is more important follow security consideration than support persistent login across multiple browsers/devices. As i said in previous message, we do not have auth keys per device to handle this correctly and in my world `security` > `persistent login in multiple devices`

**Comment by rob006** [comment]
> in my world `security` > `persistent login in multiple devices`

Then I suggest just don't enable `enableAutoLogin` at all in your applications, since this is the source of this "security problem".

**Comment by mohavee** [comment]
> > in my world `security` > `persistent login in multiple devices`
> 
> Then I suggest just don't enable `enableAutoLogin` at all in your applications, since this is the source of this "security problem".

obviously you are intentionally taking things out of context, there is no way to implement persistent login 100% secure, but atleast is good to follow best practices/considerations when yii2 ships that feature.

**Comment by rob006** [comment]
> obviously you are intentionally taking things out of context, there is no way to implement persistent login 100% secure, but atleast is good to follow best practices/considerations when yii2 ships that feature.

AFAIK the "best practice" is to not enable persistent login, so if you care about security that much, you should be consistent in it.

I repeat: this is design decision that should be made at app level and not enforced by framework. Security is not some binary state that you have or not - you should decide how much convenience you want to sacrifice to make your app more secure and you may have completely different priorities depending on whether you create forum vs payment system. 

Personally I think that this is just this kind of annoyance that make your app less secure, because it promotes bad behavior (using logout action has unwanted side effect, so people will start to avoiding it and leaving logged in sessions), same as requirement to changing password every moth (it promotes simple and predictable passwords, or just writing it on a note stuck to the monitor). Especially that it does not fix a real problem (people forgetting to logout) and does not really work because of #9718.

**Comment by mohavee** [comment]
> 
> AFAIK the "best practice" is to not enable persistent login, so if you care about security that much, you should be consistent in it.

true, but there are business requirements, which cannot be ignored and atleast i tried to implement it in best way to avoid some troubles

> I repeat: this is design decision that should be made at app level and not enforced by framework. Security is not some binary state that you have or not - you should decide how much convenience you want to sacrifice to make your app more secure and you may have completely different priorities depending on whether you create forum vs payment system.

it seems like pretty alibistic statement, framework is implementing user authentication && disclaiming that security level of feature should not be "enforced by framework"

> Personally I think that this is just this kind of annoyance that make your app less secure, because it promotes bad behavior (using logout action has unwanted side effect, so people will start to avoiding it and leaving logged in sessions), same as requirement to changing password every moth (it promotes simple and predictable passwords, or just writing it on a note stuck to the monitor). Especially that it does not fix a real problem (people forgetting to logout) and does not really work because of #9718.

ok, definitely not my behavior but whocares, its up you how to solve that problem :pray: :thumbsup:

**Comment by rob006** [comment]
> it seems like pretty alibistic statement, framework is implementing user authentication && disclaiming that security level of feature should not be "enforced by framework"

Current implementation is secure in defined scope (data channel is secure and client is not compromised) and provides basic tools to handle edge cases (you can reset auth key if you suspect that cookies might leaked). This PR does not improve security at all in case of normal environment, and does not really address issue in case of vulnerable environment, since it just triggers reset in completely unrelated action and ignores other aspects of problem (like session cookie is still valid after reseting the auth key and cases when someone does not use logout action). It is literally a fake security - you pretend that you care about stealing identity cookie, but you don't really address this problem.

**Comment by mohavee** [comment]
> > it seems like pretty alibistic statement, framework is implementing user authentication && disclaiming that security level of feature should not be "enforced by framework"
> 
> Current implementation is secure in defined scope (data channel is secure and client is not compromised) and provides basic tools to handle edge cases (you can reset auth key if you suspect that cookies might leaked). This PR does not improve security at all in case of normal environment, and does not really address issue in case of vulnerable environment, since it just triggers reset in completely unrelated action and ignores other aspects of problem (like session cookie is still valid after reseting the auth key and cases when someone does not use logout action). It is literally a fake security - you pretend that you care about stealing identity cookie, but you don't really address this problem.

so [session fixation](https://owasp.org/www-community/attacks/Session_fixation) is out of scope? if so then thanks for your time and it should be handled at app level

**Comment by rob006** [comment]
> so [session fixation](https://owasp.org/www-community/attacks/Session_fixation) is out of scope

I'm not sure how this is related to this topic.

**Comment by mohavee** [comment]
> > so [session fixation](https://owasp.org/www-community/attacks/Session_fixation) is out of scope
> 
> I'm not sure how this is related to this topic.

we currently fixate session id via identity cookie, attacker is able to obtain session id if we login by cookie, or am i wrong? its common practise to regenerate session id on login/logout and our identity cookie is tied with logged-in privilege state of user. I have met with that approach in symfony, spring auth feature as mechanism trying  to prevent attacks like session fixation/hijacking

**Comment by rob006** [comment]
Session fixation is possible only if you explicitly change PHP config - with default PHP config this attack is impossible and it is beyond me why would you want to change this. This PR also does not improve anything in this matter, since indentity cookie  does not leak in this kind of attack - attacker can obtain only session ID, which is already properly destroyed on logout action.

**Comment by mohavee** [comment]
ok, i respect your opinion if you think that stolen identity cookie is far fetched, personally ill stick to symfony/spring approach and i let app to handle it in that way

**Comment by samdark** [comment]
First of all, thanks for report, @mohavee. That triggered some digging into the topic, reading in detail about possible attacks and weighting pros and cons. After doing all that I've decided not to go this way:

1. That would be usability nightmare if you, for example, have a phone and a work computer. When you sign out of computer daily you lose session on the phone. Annoying.
2. It's more than possible to do what's done in the pull request if needed as part of your own IdentityInterface implementation.

**Pull request closed without merging**